DBW: add stylesheet gatherer and audit for display: box (old flexbox)

ebidel opened the pull request:
R: @paulirish @brendankenny @paullewis  @GoogleChrome/lighthouse 

Fixes https://github.com/GoogleChrome/lighthouse/issues/743

### Old flexbox audit

![screen shot 2016-10-11 at 3 48 03 pm](https://cloud.githubusercontent.com/assets/238208/19291458/1df9419c-8fca-11e6-9fa7-6856608515e2.png)

### Styles gatherer

Audits use the `Styles` gatherer like so:

```javascript
const styles = artifacts.Styles;
const sheetsUsingDisplayBox = StyleHelpers.stylesheetsThatUsedProperty(
    styles, 'display', 'box');
const sheetsUsingDisplay = StyleHelpers.stylesheetsThatUsedProperty(
    styles, 'display');
const sheetsUsingTranslateZHack = StyleHelpers.stylesheetsThatUsedProperty(
    styles, null, 'translateZ(0)');
```

Then they can construct a URL violation list like so:

```javascript
const urlList = [];
sheetsUsingDisplayBox.forEach(sheet => {
  sheet.parsedContent.forEach(props => {
    urlList.push({
      url: sheet.header.sourceURL,
      misc: StyleHelpers.getFormattedStyleContent(sheet.content, props)
    });
  });
});
```

brendankenny commented on lighthouse-core/config/dobetterweb.json:
```diff
@@ -49,6 +51,13 @@
     }, {
+      "name": "Using modern CSS features",
+      "audits": {
+        "uses-new-flexbox": {
```
should this be called `'no-old-flexbox'`? or something for the expected result to be false

ebidel replied:
done

brendankenny commented on lighthouse-core/gather/gatherers/styles.js:
```diff
@@ -0,0 +1,130 @@
+    this._activeStyles = [];
+    this._onStyleSheetAdded = this.onStyleSheetAdded.bind(this);
+    this._onStyleSheetRemoved = this.onStyleSheetRemoved.bind(this);
+  }
```
two overall things with this gatherer:
- need to do less work in `onStyleSheetAdded`. This is run during the page load as stylesheets come in. DBW doesn't look at traces yet, but it might at some point (and others may mix and match) so you don't want to be doing work as the page loads. Maybe the fetching and parsing of the stylesheet text be done in `endStylesCollect`?

- need to keep promise chains connected. `driver.sendCommand` returns a promise, so it should be brought in wherever possible to the existing gatherer chain (returned from the gatherer methods) so that execution ends up serialized between gatherers and so any errors are caught

ebidel replied:
done

brendankenny commented on lighthouse-core/gather/gatherers/styles.js:
```diff
@@ -0,0 +1,130 @@
+    for (let i = 0; i < this._activeStyles.length; ++i) {
+      const header = this._activeStyles[i].header;
+      if (header.styleSheetId === styleHeader.styleSheetId) {
+        this._activeStyles.splice(i, 1);
```
maybe just null these out to avoid a splice (with resulting copies, etc)? loop is slightly simpler if this is just an array of `styleSheeId`s at this point, as well

brendankenny commented on lighthouse-core/gather/gatherers/styles.js:
```diff
@@ -0,0 +1,130 @@
+  }
+
+  beginStylesCollect(opts) {
+    this.driver = opts.driver;
```
just pass in `driver` (and no need to save on `this.driver`)

ebidel replied:
done

brendankenny commented on lighthouse-core/gather/gatherers/styles.js:
```diff
@@ -0,0 +1,130 @@
+    this.driver.on('CSS.styleSheetRemoved', this._onStyleSheetRemoved);
+  }
+
+  endStylesCollect() {
```
can pass in `driver` here too (`afterPass` also gets `options`)

brendankenny commented on lighthouse-core/audits/dobetterweb/no-old-flexbox.js:
```diff
@@ -0,0 +1,83 @@
+const StyleHelpers = require('../../lib/styles-helpers');
+const Formatter = require('../../formatters/formatter');
+
+class UsesNewFlexBoxAudit extends Audit {
```
`NoOldFlexbox`? and capitalized `B` `FlexBox` or no?

ebidel replied:
done

brendankenny commented on lighthouse-core/audits/dobetterweb/no-old-flexbox.js:
```diff
@@ -0,0 +1,83 @@
+ */
+
+/**
+ * @fileoverview Audit a page to see if it is using the display: box flexbox.
```
maybe
``@fileoverview Audit a page to see if it is using the obsolete `display: box` flexbox.``
just to be clear?

ebidel replied:
done

brendankenny commented on lighthouse-core/audits/dobetterweb/no-old-flexbox.js:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    // TODO: consider usage of vendor prefixes
+    // TODO: consider usage of other older properties from
```
should track these in a DBW issue

ebidel replied:
https://github.com/GoogleChrome/lighthouse/issues/768

brendankenny commented on lighthouse-core/lib/styles-helpers.js:
```diff
@@ -0,0 +1,86 @@
+ * @param {string=} propVal Optional value of the CSS property to filter results on.
+ * @return {Array} A list of stylesheets that use the CSS property.
+ */
+function stylesheetsThatUsedProperty(stylesheets, propName, propVal) {
```
what if this only took one stylesheet at a time, and it was the caller's responsibility to loop over the stylesheet array? Could be called `stylesheetPropertyFilter` or whatever, so it was clearer it was taking in an array of styles and filtering out any that don't fit the bill.

ebidel replied:
I'm hesitant because every css audit will need to do that. We could add `stylesheetPropertyFilter` for audits to be able to call if they need it, but I think the most css audits will need to look across all sheets to find usage.

brendankenny commented on lighthouse-core/lib/styles-helpers.js:
```diff
@@ -0,0 +1,86 @@
+  });
+}
+
+/**
```
description text needed

ebidel replied:
done

brendankenny commented on lighthouse-core/lib/styles-helpers.js:
```diff
@@ -0,0 +1,86 @@
+ * @param {!Object} parsedContent Parsed CSS content.
+ * @return {string} Formatted output
+ */
+function getFormattedStyleRule(content, parsedContent) {
```
maybe this could be called in the exact same way but live on the `UrlList` formatter class, since it's mostly concerned with formatting the rule for the formatter to print?

ebidel replied:
It's unrelated to URLs though. It's formatting a snippet of css content.

brendankenny replied:
ha, but you're the one using it for every DBW extended info :P You could rename it to the `GenericDBWList` formatter

brendankenny commented on lighthouse-core/lib/styles-helpers.js:
```diff
@@ -0,0 +1,86 @@
+  return block;
+}
+
+module.exports = {
```
`stylesheetsThatUsedProperty` needs some tests

ebidel replied:
done

ebidel commented on lighthouse-core/gather/gatherers/styles.js:
```diff
@@ -0,0 +1,130 @@
+const Gatherer = require('./gatherer');
+
+/**
+ * @param {!Artifacts} parseTree
```
done

ebidel commented on lighthouse-core/gather/gatherers/styles.js:
```diff
@@ -0,0 +1,130 @@
+        const map = new Map(stylesheets.map(s => [s.content, s]));
+        this.artifact = Array.from(map.values());
+      }, _ => {
+        this.artifact = -1;
```
done

ebidel commented:
PTAL

paulirish commented on lighthouse-core/gather/gatherers/styles.js:
```diff
@@ -0,0 +1,143 @@
+  let results = [];
+
+  parseTree.traverseByType('declaration', function(node, index, parent) {
+    const [name, val] = node.toString().split(':').map(item => item.trim());
```
We've avoided destructuring up till now. AFAIK it's not supported back to node 4 --harmony. Can we do this without destructuring?

ebidel replied:
done

paulirish commented on lighthouse-core/gather/gatherers/styles.js:
```diff
@@ -0,0 +1,143 @@
+      Promise.all(contentPromises).then(_ => {
+        driver.off('CSS.styleSheetAdded', this._onStyleSheetAdded);
+        driver.off('CSS.styleSheetRemoved', this._onStyleSheetRemoved);
+        driver.sendCommand('CSS.disable');
```
DOM.disable?

ebidel replied:
done

paulirish commented on lighthouse-core/gather/gatherers/styles.js:
```diff
@@ -0,0 +1,143 @@
+          const styleHeader = this._activeStyleHeaders[sheetId];
+          styleHeader.content = content.text;
+          styleHeader.parsedContent = getCSSPropsInStyleSheet(
+              parser.parse(styleHeader.content, {syntax: 'css'}));
```
parse() takes a 2nd arg?

ebidel replied:
yep, https://github.com/tonyganch/gonzales-pe/blob/dev/src/parse.js#L15

paulirish replied:
you call that one via your own custom ` WebInspector.SCSSParser.prototype.parse = function(content) {`.. and its the SCSSParser instance that you're calling here. 

not a big deal, just noticing.

ebidel commented:
PTA👀

ebidel commented:
OH yay. Travis is happy now. 

Safari TP fixed a bunch of Shadow DOM bugs and my coffee is extra strong this morning....Today is going to be a good day!

brendankenny commented on lighthouse-core/gather/gatherers/styles.js:
```diff
@@ -0,0 +1,144 @@
+        });
+      });
+
+      Promise.all(contentPromises).then(_ => {
```
the results of `contentPromises` is the input to this `then` function, so no need to reconstruct from `Object.keys(this._activeStyleHeaders)`. Also lets you bring in `sendCommand` calls. Something like

```js
Promise.all(contentPromises).then(stylesheets => {
  driver.off('CSS.styleSheetAdded', this._onStyleSheetAdded);
  driver.off('CSS.styleSheetRemoved', this._onStyleSheetRemoved);
  return driver.sendCommand('CSS.disable')
    .then(_ => driver.sendCommand('DOM.disable'))
    .then(_ => resolve(stylesheets));
}).catch(err => reject(err));
```
(or one of the many promise chaining permutations that could be done with this)

ebidel replied:
done

brendankenny commented on lighthouse-core/lib/styles-helpers.js:
```diff
@@ -0,0 +1,101 @@
+ * @param {!Object} parsedContent The parsed version content.
+ * @return {string} Formatted output
+ */
+function getFormattedStyleRule(content, parsedContent) {
```
still think this should live on the formatter. This is purely a function for formatting for the report printer

ebidel replied:
going to leave as is for now.

paulirish replied:
+1 keeping it as is. I would prefer heavier work inside /lib/ and formatters are fairly dumb. In the future i'd be into fixing the critical-network-chain formatters to be more like that.

brendankenny commented on lighthouse-core/test/audits/dobetterweb/no-old-flexbox-test.js:
```diff
@@ -0,0 +1,49 @@
+/* eslint-env mocha */
+
+describe('Page does not use old CSS flexbox', () => {
+  it('fails when no input present', () => {
```
maybe `it returns error value when no input present`? or something to differentiate "fail" due to an error vs "fail" due to having old flexbox

ebidel replied:
done

brendankenny commented on lighthouse-core/test/audits/dobetterweb/no-old-flexbox-test.js:
```diff
@@ -0,0 +1,49 @@
+    assert.equal(auditResult.rawValue, -1);
+    assert.ok(auditResult.debugString);
+  });
+
```
maybe add a test that checks a `debugString` from the gatherer (passed in on the artifacts object) makes it through the audit? Not sure how specific you want to get on that, though.

ebidel replied:
done

brendankenny commented on lighthouse-core/test/lib/style-helpers-test.js:
```diff
@@ -0,0 +1,76 @@
+      assert.equal(results.length, 1, 'accepts CSS property name/value pair');
+
+      results = StyleHelpers.filterStylesheetsByUsage(
+          stylesheets, 'display');
```
maybe change this to a property name that doesn't work with the value below, just to ensure that `filterStylesheetsByUsage` isn't changing the original stylesheets? Maybe use `color` here, so it's filtered only for `color` entries, for example, but keep `box` below, which will never be the value for a `color` property?

brendankenny commented on lighthouse-core/test/lib/style-helpers-test.js:
```diff
@@ -0,0 +1,76 @@
+    it('returns no results when not found', () => {
+      let results = StyleHelpers.filterStylesheetsByUsage(
+          stylesheets, 'someunknownprop', 'someunknownval');
+      assert.equal(results.length, 0, 'CSS property name/value pair not found');
```
maybe add an assertion for a known prop and an unknown value (or vice versa)

ebidel replied:
done

ebidel commented on lighthouse-core/gather/gatherers/styles.js:
```diff
@@ -0,0 +1,144 @@
+    driver.sendCommand('DOM.enable');
+    driver.sendCommand('CSS.enable');
+    driver.on('CSS.styleSheetAdded', this._onStyleSheetAdded);
+    driver.on('CSS.styleSheetRemoved', this._onStyleSheetRemoved);
```
donedone

ebidel commented on lighthouse-core/gather/gatherers/styles.js:
```diff
@@ -0,0 +1,144 @@
+  }
+
+  beforePass(options) {
+    this.beginStylesCollect(options.driver);
```
done

ebidel commented on lighthouse-core/gather/gatherers/styles.js:
```diff
@@ -0,0 +1,144 @@
+          rawValue: -1,
+          debugString: err
+        };
+        return;
```
done

ebidel commented:
Think that's it.

brendankenny commented on lighthouse-core/audits/dobetterweb/no-old-flexbox.js:
```diff
@@ -0,0 +1,86 @@
+   */
+  static audit(artifacts) {
+    if (typeof artifacts.Styles === 'undefined' ||
+        artifacts.Styles === -1) {
```
remove `artifacts.Styles === -1` case

brendankenny commented on lighthouse-core/gather/gatherers/styles.js:
```diff
@@ -0,0 +1,144 @@
+
+  endStylesCollect(driver) {
+    return new Promise((resolve, reject) => {
+      if (!driver || !this._activeStyleSheetIds.length) {
```
you can remove `!driver` case

brendankenny commented on lighthouse-core/lib/styles-helpers.js:
```diff
@@ -0,0 +1,101 @@
+ *     on. If propVal is not specified, all stylesheets that use the property are
+ *     returned. Otherwise, stylesheets that use the propName: propVal are returned.
+ * @param {string=} propVal Optional value of the CSS property to filter results on.
+ * @return {Array} A list of stylesheets that use the CSS property.
```
`{!Array}`

ebidel replied:
done

brendankenny commented on lighthouse-core/lib/styles-helpers.js:
```diff
@@ -0,0 +1,101 @@
+    const endLine = lines[declarationRange.endLine];
+    rule = lines.slice(startLine, endLine).reduce((prev, line) => {
+      prev.push(line.substring(
+          declarationRange.startColumn, declarationRange.endColumn));
```
replace with `start`/`end`?

ebidel replied:
done

brendankenny commented on lighthouse-core/test/audits/dobetterweb/no-old-flexbox-test.js:
```diff
@@ -0,0 +1,60 @@
+      }
+    });
+    assert.equal(auditResult.rawValue, -1);
+    assert.ok(auditResult.debugString);
```
check it's equal to what was passed in?

ebidel replied:
done

paulirish commented:
lgtm here. thx

ebidel commented:
Don't merge yet. I'm fixing something related to multiple values.

ebidel commented:
I fixed the bug with multi css property values. The output now looks something like this:

![screen shot 2016-10-12 at 6 11 11 pm](https://cloud.githubusercontent.com/assets/238208/19333180/5e8b3fd8-90a8-11e6-9979-3ad63fcec5bd.png)

As part of this, I changed `protocol` to a more generic `label` in the url formatter and moved row/line/col violations to be alongside URLs. That's why there are more files touched in this PR. Using `protocol` didn't make sense anymore b/c `uses-http2` audit was the only one to use the protocol.

brendankenny commented on lighthouse-core/audits/dobetterweb/uses-http2.js:
```diff
@@ -63,6 +63,11 @@ class UsesHTTP2Audit extends Audit {
+    }).map(record => {
+      return Object.assign({
+        label: record.protocol,
+        url: record.url
```
isn't url already copied over by the `Object.assign()`?

ebidel replied:
added a comment on the getter not being copied over.

brendankenny commented on lighthouse-core/lib/styles-helpers.js:
```diff
@@ -55,21 +55,17 @@ function filterStylesheetsByUsage(stylesheets, propName, propVal) {
  * @param {!string} content CSS text content.
  * @param {!Object} parsedContent The parsed version content.
- * @return {string} Formatted output
+ * @return {!Object} Formatted output.
```
`@return {{styleRule: string, location: string}} Formatted output.`

ebidel replied:
done
